<a href="https://colab.research.google.com/github/Shahrukh2016/Logistic_Regression_Implementation/blob/main/LogisticRegression_Gender_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Hi! this is Kaggle dataset obtained from Kaggle.com. I've used this dataset for building the classification Machine Learning Model.**

**It contains 7 features and a label column.**

**longhair** - This column contains 0's and 1's where 1 is "long hair" and 0 is "not long hair".

**foreheadwidthcm** - This column is in CM's. This is the width of the forehead.

**foreheadheightcm** - This is the height of the forehead and it's in Cm's.

**nosewide** - This column contains 0's and 1's where 1 is "wide nose" and 0 is "not wide nose".

**noselong** - This column contains 0's and 1's where 1 is "Long nose" and 0 is "not long nose".

**lipsthin** - This column contains 0's and 1's where 1 represents the "thin lips" while 0 is "Not thin lips".

**distancenosetoliplong** - This column contains 0's and 1's where 1 represents the "long distance between nose and lips" while 0 is "short distance between nose and lips".

**gender** - This is either "Male" or "Female".

"Gender is our target variable"

##**Data Pre-Processing**

In [43]:
# Importing fundamental essentical libraries
import numpy as np
import pandas as pd

In [44]:
# Reading our dataset
df= pd.read_csv("/content/drive/MyDrive/Colab Notebooks/ML Algorithms Shahrukh/Datasets/Gender_Classification.csv")

In [45]:
df.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1    Male  
1          1                          0  Female  
2          1                          1    Male  
3          1                          1    Male  
4          0                          0  Female

First thing first. Lets check our dataset if it has any null, missing, duplicate obsrvations and treat them accordingly.

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   long_hair                  5001 non-null   int64  
 1   forehead_width_cm          5001 non-null   float64
 2   forehead_height_cm         5001 non-null   float64
 3   nose_wide                  5001 non-null   int64  
 4   nose_long                  5001 non-null   int64  
 5   lips_thin                  5001 non-null   int64  
 6   distance_nose_to_lip_long  5001 non-null   int64  
 7   gender                     5001 non-null   object 
dtypes: float64(2), int64(5), object(1)
memory usage: 312.7+ KB


No null values observed.

In [47]:
# checking duplicated observations
df.duplicated().sum()

1768

In [48]:
# total number of observations
df.shape[0]

5001

Out of 5001 observations we have 1768 duplicated values that makes no sense and will generate unnecessary noice in our predictions. Let's drop them first.


In [49]:
#dropping duplicated observations
df.drop_duplicates(inplace=True)

In [50]:
# rechecking duplicated observations
df.duplicated().sum()

0

Hurray!!! We do not have any null or duplicated value now.

Since this is classification dataset(output variable is discrete/catagorical-Male/Female), to develop and deploy any ML algorithm first we need to format our string variabel to numeric. But before doing that let's just ceate a copy of our dataframe to avoid any blunder in our original dataframe.

In [51]:
#creating copy of our dataframe
df_new= df.copy()

In [52]:
df_new.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1    Male  
1          1                          0  Female  
2          1                          1    Male  
3          1                          1    Male  
4          0                          0  Female

To make our output variable numeric so that it can be trained by any ML Algorithm, let's follow up with the LABEL ENCODING.

###**Label Encoding**

In [53]:
#Formatting Male as 1 and Female as 0
# Importing required library
from sklearn.preprocessing import LabelEncoder

#create instance of label encoder
lab = LabelEncoder()

#perform label encoding on 'team' column
df_new['gender'] = lab.fit_transform(df_new['gender'])

In [54]:
#checking encoded column
df_new.head()

long_hair  forehead_width_cm  forehead_height_cm  nose_wide  nose_long  \
0          1               11.8                 6.1          1          0   
1          0               14.0                 5.4          0          0   
2          0               11.8                 6.3          1          1   
3          0               14.4                 6.1          0          1   
4          1               13.5                 5.9          0          0   

   lips_thin  distance_nose_to_lip_long  gender  
0          1                          1       1  
1          1                          0       0  
2          1                          1       1  
3          1                          1       1  
4          0                          0       0

For the sake of simplicity let's just take our input variable as "long_hair" and output variable as "gender" of course and make the predictions after that we will consider more features and will see if we are gaining more accuracy in our model. 

###**Train test split**

In [55]:
#Assigning observations to our x and y variable
x= df_new[["long_hair"]]
y= df_new[["gender"]]

In [56]:
# checking shape of dependent and independent variable
x.shape, y.shape

((3233, 1), (3233, 1))

In [57]:
# Importing sklearn library for splitting our data in train and test observations
from sklearn.model_selection import train_test_split

In [58]:
#split operation
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [59]:
x_train

long_hair
380           1
137           1
2357          1
3215          0
1433          1
...         ...
1282          1
1328          0
1555          1
974           1
4874          0

[2586 rows x 1 columns]

In [60]:
y_train

gender
380        1
137        0
2357       1
3215       1
1433       0
...      ...
1282       1
1328       0
1555       0
974        0
4874       1

[2586 rows x 1 columns]

In [61]:
x_test

long_hair
1187          0
1688          1
1470          1
30            1
141           1
...         ...
765           1
3694          1
900           0
1333          1
1655          1

[647 rows x 1 columns]

In [62]:
y_test

gender
1187       1
1688       0
1470       0
30         1
141        1
...      ...
765        0
3694       0
900        0
1333       1
1655       1

[647 rows x 1 columns]

In [63]:
#checking shape of our splitted dataset
x_train.shape, x_test.shape, y_train.shape, y_test.shape

((2586, 1), (647, 1), (2586, 1), (647, 1))

In [64]:
# checking the distribution ratio of each of catagory in our train dataset 
y_train.value_counts()[1]/y_train.value_counts()[0]

1.2644483362521892

In [65]:
# checking the distribution ratio of each of catagory in our test dataset 
y_test.value_counts()[1]/y_test.value_counts()[0]

1.1006493506493507

Since our catagory distribution ratio is nearly same that means we have balanced dataset in our train and test samples which is good for us. 

##**Model Implementation and Validation**

###**Model Training**

In [66]:
# importing LogisticRegression from sklearn
from sklearn.linear_model import LogisticRegression

In [67]:
# creating instance
classification= LogisticRegression(fit_intercept=True,max_iter=1000000)

In [68]:
#Training the model
classification.fit(x_train,y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(max_iter=1000000)

In [69]:
#checking scoring parameter for our trainins dataset
classification.score(x_train, y_train)

0.5583913379737045

In [70]:
#checing the values of cofficient and intercept
classification.coef_, classification.intercept_

(array([[0.10687628]]), array([0.14710901]))

###**Model Testing**

In [71]:
#output train dataset prediction
y_pred_train= classification.predict(x_train)
print(y_pred_train)

[1 1 1 ... 1 1 1]


In [72]:
#output test dataset prediction
y_pred_test= classification.predict(x_test)
print(y_pred_test)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 

In [73]:
# calculating the probabilities of outcume for our training dataset
y_pred_prob_train= classification.predict_proba(x_train)
print(y_pred_prob_train)

[[0.43684283 0.56315717]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 ...
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 [0.46328893 0.53671107]]


In [74]:
# calculating the probabilities of outcume for our testing dataset
y_pred_prob_test= classification.predict_proba(x_test)
print(y_pred_prob_test)

[[0.46328893 0.53671107]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]
 ...
 [0.46328893 0.53671107]
 [0.43684283 0.56315717]
 [0.43684283 0.56315717]]


###**Model Validation**

In [75]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import log_loss
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import roc_auc_score

In [77]:
# checking different accuracy metrics for training dataset
print(f"Confusion Matrix is: {confusion_matrix(y_train,y_pred_train)}")
print(f"Accuracy Score is: {accuracy_score(y_train,y_pred_train)*100} %")
print(f"Log loss is: {log_loss(y_train,y_pred_train)}")
print(f"Precision Score is: {precision_score(y_train,y_pred_train)*100} %")
print(f"Recall Score is: {recall_score(y_train,y_pred_train)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_train,y_pred_train)}")

Confusion Matrix is: [[   0 1142]
 [   0 1444]]
Accuracy Score is: 55.83913379737046 %
Log loss is: 15.25297594093321
Precision Score is: 55.83913379737046 %
Recall Score is: 100.0 %
ROC_AUC Score is: 0.5


In [78]:
# checking different accuracy metrics for testing dataset
print(f"Confusion Matrix is: {confusion_matrix(y_test,y_pred_test)}")
print(f"Accuracy Score is: {accuracy_score(y_test,y_pred_test)*100} %")
print(f"Log loss is: {log_loss(y_test,y_pred_test)}")
print(f"Precision Score is: {precision_score(y_test,y_pred_test)*100} %")
print(f"Recall Score is: {recall_score(y_test,y_pred_test)*100} %")
print(f"ROC_AUC Score is: {roc_auc_score(y_test,y_pred_test)}")

Confusion Matrix is: [[  0 308]
 [  0 339]]
Accuracy Score is: 52.39567233384853 %
Log loss is: 16.44233292989334
Precision Score is: 52.39567233384853 %
Recall Score is: 100.0 %
ROC_AUC Score is: 0.5
